NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,467 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+	virtual void Run( int threadIndex, int index, int count );
+	virtual void SubRun( int /*threadIndex*/, int /*index*/ ) { NeoAssert( false ); }
+
+	const int VectorSize;
+	const CKMeansClustering& Owner;
+};
+
+void CKMeansClustering::IThreadTask::CallRun( int threadIndex, int align )
+{
+	int index = 0;
+	int count = 0;
+	if( GetTaskIndexAndCount( ThreadCount(), threadIndex, Size(), align, index, count ) ) {
+		Run( threadIndex, index, count );
+	}
+}
+
+void CKMeansClustering::IThreadTask::Run( int threadIndex, int index, int count )
+{
+	const int last = index + count;
+	for( int i = index; i < last; ++i ) {
+		SubRun( threadIndex, i );
+	}
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CClassifyAllThreadTask final : public CKMeansClustering::IThreadTask {
+	CClassifyAllThreadTask( const CKMeansClustering& owner, const CFloatMatrixDesc& matrix ) :
+		IThreadTask( owner, &matrix ),
+		Inertia( ThreadCount() )
+	{ DataCluster.SetBufferSize( Size() ); }
+
+	CArray<int> DataCluster{}; // the cluster for this element
+	CInertia Inertia;
+
+protected:
+	void SubRun( int threadIndex, int index ) override
+	{ DataCluster[index] = Owner.findNearestCluster( *Matrix, index, Inertia.Get( threadIndex ) ); }
+};
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CAssignVectorsThreadTask final : public CKMeansClustering::IThreadTask {
+	CAssignVectorsThreadTask( const CKMeansClustering& owner, const CFloatMatrixDesc& matrix ) :
+		IThreadTask( owner, &matrix )
+	{}
+
+	// Element assignments (objectCount)
+	CArray<int> Assignments{};
+	// Distances bounds
+	CArray<float> UpperBounds{}; // upper bounds for every object (objectCount)
+	CVariableMatrix<float> LowerBounds{}; // lower bounds for every object and every cluster (clusterCount x objectCount)
+
+	// Distances between clusters (clusterCount x clusteCount)
+	CVariableMatrix<float> ClusterDists{};
+	// Distance to the closest center of another cluster (clusterCount)
+	CArray<float> ClosestClusterDist{};
+	// Distances between old and updated centers of each cluster (clusterCount)
+	CArray<float> MoveDistance{};
+
+protected:
+	void SubRun( int threadIndex, int index ) override;
+	// Checks if operation can be omitted
+	// If it's true element 'id' can't be reassigned to the cluster 'clusterToProcess'
+	// and all of the calculations related to this case can be skipped
+	bool isPruned( int clusterToProcess, int id ) const;
+};
+
+bool CKMeansClustering::CAssignVectorsThreadTask::isPruned( int clusterToProcess, int id ) const
+{
+	return ( Assignments[id] == clusterToProcess ) ||
+		( UpperBounds[id] <= LowerBounds( clusterToProcess, id ) ) ||
+		( UpperBounds[id] <= 0.5 * ClusterDists( Assignments[id], clusterToProcess ) );
+}
+
+void CKMeansClustering::CAssignVectorsThreadTask::SubRun( int /*threadIndex*/, int i )
```

Comment:
`int i` outside of a small cycle? Better something more sensible... Like `vectorIndex`

Reply:
Code inside originally already had that name, I economized renaming :)
I will rename)

Reply:
Okay, anything which makes sense. One-letter is cool inside small circles and in some specific cases where it's a common thing (like some math code).